File: src/levels/L04-singleton/ts/factory.ts
class CounterFactory {
  // region: CounterFactory#create
  create() { return new TicketCounter() } // un contador nuevo en cada pedido: siempre turno 1
  // endregion
}

class TicketCounter {
  private last = 0
  next() { return ++this.last }
}

class Register {
  constructor(private display: Display) {}

  // region: Register#checkout
  checkout(order: Order) {
    order.ticket = new CounterFactory().create().next()
    this.display.show(order)
  }
  // endregion
}

File: src/levels/L04-singleton/factory.rb
class CounterFactory
  # region: CounterFactory#create
  def create = TicketCounter.new # un contador nuevo en cada pedido: siempre turno 1
  # endregion
end

class TicketCounter
  def initialize = @last = 0
  def next = @last += 1
end

class Register
  def initialize(display) = @display = display

  # region: Register#checkout
  def checkout(order)
    order.ticket = CounterFactory.new.create.next
    @display.show(order)
  end
  # endregion
end
